File: src/Solutions/Euler116.hs
module Solutions.Euler116 where
import Data.Array.Unboxed
import Data.Array.ST
import Control.Monad

generator :: Int -> UArray Int Int
generator k = runSTUArray $ do
    arr <- newArray_ (0, 50)
    forM_ [0..k - 1] $ \x -> writeArray arr x 1
    forM_ [k..50] $ \x -> (+) <$> readArray arr (x - 1) <*> readArray arr (x - k) >>= writeArray arr x
    mapArray pred arr -- remove option with grey tiles only

red = generator 2
green = generator 3
blue = generator 4

euler116 :: IO String
euler116 = return $ show $ red ! 50 + blue ! 50 + green ! 50

File: rust/src/solutions/euler116.rs
fn generator(k: usize) -> Vec<u64> {
    let mut vec = vec![1u64; 51];
    for i in k..51 {
        vec[i] = vec[i - 1] + vec[i - k];
    }
    for i in 0..51 {
        vec[i] -= 1;
    }
    vec
}

pub fn euler116() -> String {
    (generator(2)[50] + generator(3)[50] + generator(4)[50]).to_string()
}
